### PR TITLE
docs(api): begin work on stats json documentation

### LINKED ISSUE #491: API - Stats JSON Schema
I'm working on some tools to help my company find areas to optimize our builds. I'm doing this primarily through traversing the stats file. There's documentation on the options I can toggle on my builds, but I haven't found any good descriptions of the structure of the json itself. For the most part it seems pretty straight forward, I imagine reasons is a list of modules that include a given module, but I'm seeing some unexpected results when I try to use moduleId and the modules array to filter reasons by chunk. It would be very helpful to have a reference for this information. Does this exist anywhere? If so, can it be made public?

### PR DESCRIPTION
**Pull request by skipjack**
Add page to document the output of...

``` bash
webpack --profile --json > compilation-stats.json
```

We still need to go through and fill in the blanks before this can be merged.

Resolves #491 

cc @dmitriid @walfly

### CODE REVIEW

File: content/api/stats.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+
+### Module Objects
+
+What good would these statistics be without some description of the compiled application's actual modules? Each module in the dependency graph is represented by the following structure:
+
+``` js
+{
+  "assets": [], // A list of [asset objects](#asset-objects) (?)
+  "built": true, // Maybe this means compiled, loaded or output (?)
```
Comment (sokra) [inline content/api/stats.md]: Loaders, Parsing, Code Generation

File: content/api/stats.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  "cacheable": true, // Whether or not this module is cacheable (?)
+  "chunks": [
+    // IDs of chunks that contain this module (?)
+  ],
+  "errors": 0, // Number of errors when resolving or processing the module (?)
+  "failed": false, // Whether or not compilation failed on this module (?)
+  "id": 0, // The ID of the module
```
Comment (sokra) [inline content/api/stats.md]: By default these ids for modules and chunks are numeric, but they could be strings too (NamedModulesPlugin, NamedChunksPlugins)

File: content/api/stats.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  "identifier": "(webpack)\\test\\browsertest\\lib\\index.web.js", // Another ID ?
+  "name": "./lib/index.web.js", // Path to the actual file
+  "optional": false, // ?
```
Comment (sokra) [inline content/api/stats.md]: All requires to this module are inside of `try-catch`-blocks. Pretty much irrelevant with ESM.

File: content/api/stats.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  "prefetched": false, // Whether or not the module was prefetched (a build optimization)
+  "profile": {
+    // Module specific compilation stats corresponding to the [`--profile` flag](/api/cli#profiling)
+    "building": 73,
+    "dependencies": 242,
+    "factory": 11
```
Comment (sokra) [inline content/api/stats.md]: factory -> resolving

File: content/api/stats.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  "errors": 0, // Number of errors when resolving or processing the module (?)
+  "failed": false, // Whether or not compilation failed on this module (?)
+  "id": 0, // The ID of the module
+  "identifier": "(webpack)\\test\\browsertest\\lib\\index.web.js", // Another ID ?
+  "name": "./lib/index.web.js", // Path to the actual file
+  "optional": false, // ?
+  "prefetched": false, // Whether or not the module was prefetched (a build optimization)
+  "profile": {
+    // Module specific compilation stats corresponding to the [`--profile` flag](/api/cli#profiling)
+    "building": 73,
```
Comment (sokra) [inline content/api/stats.md]: loaders + parsing

File: content/api/stats.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    "dependencies": 242,
```
Comment (sokra) [inline content/api/stats.md]: how long it took until all dependency of this module finished

File: content/api/stats.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    "factory": 11
+  },
+  "reasons": [
+    // See the description below...
+  ],
+  "size": 3593, // Module size in bytes?
```
Comment (sokra) [inline content/api/stats.md]: This size is only estimated.

File: content/api/stats.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  "assets": [], // A list of [asset objects](#asset-objects) (?)
+  "built": true, // Maybe this means compiled, loaded or output (?)
+  "cacheable": true, // Whether or not this module is cacheable (?)
+  "chunks": [
+    // IDs of chunks that contain this module (?)
+  ],
+  "errors": 0, // Number of errors when resolving or processing the module (?)
+  "failed": false, // Whether or not compilation failed on this module (?)
+  "id": 0, // The ID of the module
+  "identifier": "(webpack)\\test\\browsertest\\lib\\index.web.js", // Another ID ?
```
Comment (sokra) [inline content/api/stats.md]: identifier is a unique internal id.

id is the final id that is emitted and exposed.

File: content/api/stats.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+Every module also contains a list of `reasons` objects describing why that module was included in the dependency graph. Each "reason" is very similar to the `origins` objects seen above in the [chunk objects](#chunk-objects) section:
+
+``` js
+{
+  "loc": "33:24-93", // Lines of code that caused the module to be included
+  "module": "./lib/index.web.js", // ?
+  "moduleId": 0, // The ID of the module
+  "moduleIdentifier": "(webpack)\\test\\browsertest\\lib\\index.web.js", // ?
+  "moduleName": "./lib/index.web.js", // ?
```
Comment (sokra) [inline content/api/stats.md]: This is a pretty-printed/**readable** name like it's disabled in the string stats.

File: content/api/stats.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  "type": "require.context", // The [type of request](/api/module-methods) used
+  "userRequest": "../../cases" // ?
```
Comment (sokra) [inline content/api/stats.md]: The raw string that the user used inside of the `require()` or `import`.

File: content/api/stats.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+}
+```
+
+
+### Errors and Warnings
+
+The `errors` and `warnings` properties each contain a list of strings. Each string contains a message and stack trace:
```
Comment (sokra) [inline content/api/stats.md]: stacktrace is only included when `errorDetails: true` was passed to `toJson` (that the default)

File: content/api/stats.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+
+## Structure
+
+``` js-with-links
+{
+  "version": "1.4.13", // Version of webpack used for the compilation
+  "hash": "11593e3b3ac85436984a", // Compilation specific hash
+  "time": 2469, // Compilation time in milliseconds (?)
+  "filteredModules": 0, // ?
```
Comment (skipjack) [inline content/api/stats.md]: 👍

**Comment by skipjack** [comment]
@sokra should be ready for another pass. I filled in some other bits as well and removed all the question marks you didn't touch on (assumed those points were ok). Let me know if there's anything else that needs updating.

@walfly @dmitriid any thoughts? This probably isn't perfect 👌  but it should be a good start for us to build on... if anything else arises we can tackle it with more targeted prs/issues.

**Comment by dmitriid** [comment]
This is absolutely wonderful. Thank you!

File: content/api/stats.md (hunk 11 above)
Comment (skipjack) [inline content/api/stats.md]: Noted below (line 178).

**Comment by skipjack** [comment]
Ok even though I don't think all the descriptions are perfect, I'm going to merge as this is definitely a step up from _no documentation_. @sokra @walfly @dmitriid please submit prs or issues with any clarifications or improvements (or add suggestions here).

**Pull request merged**